### PR TITLE
Docs for communication API and updates for provider plugin type - MDL-79961

### PR DESCRIPTION
**Pull request by safatshahin**
This doc is related to Communication API and provider plugin type.

To track the progress of this task, I used MDL-79961 in our current sprint.

### CODE REVIEW

**Comment by sarjona** [comment]
@safatshahin is this pull-request blocked by https://tracker.moodle.org/browse/MDL-78551 (so should wait for this issue to be integrated before merging it)?

Apart from that, it would be great if you could fix the failures ;-)

**Comment by safatshahin** [comment]
Hi @sarjona 
I have removed the dependency of MDL-78551 and also moved the docs specifically related to that tracker to a different branch, I will push that later while 78551 is in integration, and we can move this forward as we don't have comm API docs and the plugin doc is not up to date.
I have resolved the CI errors, please give it another go and you can remove the blocked-by-tracker label.
Cheers!

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:27]: I think it would be better to give the exact name of this experimental setting, to help admins find it, something like Enable communication providers (enablecommunicationsubsystem).

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:1]: There is a typo in the commit message: Communicaiton

Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:1]: I think the changes on this pull request should be backported to the versioned_docs/version-4.3 branch (or at least backport the ones that apply to Moodle 4.3).

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:33]:
```suggestion
To load a communication instance, you must use the `communication\api::load_by_instance()` method. This method will return a communication instance which can be
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:35]:
```suggestion
context is the actual course context object, component is the component name (`core_course`), instance type is the custom string which can change according to
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:37]:
```suggestion
the `load_by_instance()` method. The provider is the name of the provider plugin which is used to load the instance. It's optional, if not provided, it will load
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:56]:
```suggestion
`$communication->create_and_configure_room()` method is used to add an ad-hoc to create a room in the communication provider and configure it with the required settings.
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:78]:
```suggestion
// First initialize the communication instance, provided the context is already initialized and the selected provider is available
// though a form or other means.
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:87]:
```suggestion
// Now call the create_and_configure_room() method to add an ad-hoc to create a room in the communication provider and configure it
// with the required settings.
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [reply docs/apis/subsystems/communication/index.md:78]: The given examples should follow the coding style rules (because they will be taken as a good practices to follow) ;-)

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:97]:
```suggestion
`$communication->update_room()` method is used to add an ad-hoc to update a room in the communication provider and configure it with the required settings.
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:125]: This probably also needs to be reviewed (the comment should be split in different lines).

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:136]:
```suggestion
`$communication->delete_room()` method is used to add an ad-hoc to delete a room in the communication provider. The associated task for this also removes all the members from
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:141]:
```suggestion
`$communication->add_members_to_room()` method is used to add an ad-hoc to add members to a room in the communication provider. The user id of each user to be added to the provider room is also stored in
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:173]:
```suggestion
`$communication->update_room_membership()` method is used to add an ad-hoc to update the membership of a room in the communication provider. The user mapping for each of the users is also reset to allow task to update the users from the task.
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:182]:
```suggestion
This method accepts the same parameters as the `add_members_to_room()` method and the usage is also the same.
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:186]:
```suggestion
`$communication->remove_members_from_room()` method is used to add an ad-hoc to remove members from a room in the communication provider. The users are flagged as deleted in the `communication_user` table to allow the task to remove the users from the provider room.
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:137]:
```suggestion
the room before deleting the room.

:::danger

This is destructive and might remove all the messages and other data associated with the room. Usage of this should be done with caution.

:::

```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:195]:
```suggestion
This method accepts the same parameters as the `add_members_to_room()` method and the usage is also the same.
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:199]:
```suggestion
`$communication->show_communication_room_status_notification()` is a special method to have a UI element to show the status of the room creation. If the room is ready, it will
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:31]:
```suggestion
## Loading a communication instance
```

Comment (sarjona) [reply docs/apis/subsystems/communication/index.md:31]: The first level should be ## so all the headings in this file should be reviewed too.

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:205]:
```suggestion
The communication API allows settings to configure the basic information like room name, selection of provider etc. `$communication->form_definition()` method
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:206]:
```suggestion
is used to get the form definition for the settings form. If the form elements are used in another form to include the communication form elements, `form_definition()` method
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:207]:
```suggestion
will be useful. There is a configuration page (`communication/configure.php`) which can be used to configure the communication instance. Course currently uses this page to set up
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:209]:
```suggestion
It will also be required to use `set_data()` method in order to set the data to the form elements while saving the data.
```

File: docs/apis/subsystems/communication/index.md
Comment (sarjona) [inline docs/apis/subsystems/communication/index.md:244]:
```suggestion
support. 


:::info

Please refer to [Communication plugin](../../plugintypes/communication/index.md) documentation for more details about building a plugin.

:::
```

File: docs/apis/plugintypes/communication/index.md
Comment (sarjona) [inline docs/apis/plugintypes/communication/index.md:76]:
```suggestion
on which interfaces should be implemented, plugins can choose which features they support and implement accordingly. Exception is the `communication_provider` interface which
```

File: docs/apis/plugintypes/communication/index.md
Comment (sarjona) [inline docs/apis/plugintypes/communication/index.md:105]:
```suggestion
via API and the `communication_matrix` plugin can support the creation of users in Matrix, in that case, `communication_matrix` plugin should implement this interface. Some APIs might
```

File: docs/apis/plugintypes/communication/index.md
Comment (sarjona) [inline docs/apis/plugintypes/communication/index.md:154]:
```suggestion
Communication API has a scheduled task `core_communication\task\synchronise_providers_task` which will synchronise the data from the communication provider to the current Moodle
```

**Review by sarjona — CHANGES_REQUESTED** [review]
Hi @safatshahin!
Thanks for clarifying and updating the dependencies :-)
The docs look good. However, as you'll see, I've raised a few points that should be reviewed ;-)
If you have any question, let me know.

File: docs/apis/subsystems/communication/index.md
Comment (safatshahin) [reply docs/apis/subsystems/communication/index.md:27]: Done, added the steps for better understanding.

File: docs/apis/subsystems/communication/index.md
Comment (safatshahin) [reply docs/apis/subsystems/communication/index.md:33]: done

File: docs/apis/subsystems/communication/index.md
Comment (safatshahin) [reply docs/apis/subsystems/communication/index.md:35]: done

**Comment by safatshahin** [comment]
Hi @sarjona 
Thank you for your comments. I have addressed all of them, added the 4.3 forked version for the subsystem, and updated the 4.3 docs for plugin type. The current doc will not need any changes for 4.3 as the changes related to 4.4 are done in a different branch.

I have 4 commits to make changes in 4 different files which should make it easier to review.

Please give it another go and let me know of any other changes.

Cheers!

**Comment by safatshahin** [comment]
Hi @sarjona 
The tracker is closed so that the review can continue in the GH issue.
Please let me know if you want any other changes.
Cheers!

**Review by sarjona — APPROVED** [review]
Hi @safatshahin!
Thanks a lot for reviewing the raised points. The docs look good so I'm happy to approve this pull-request :-)

**Pull request merged**
